Port securesystemslib.hash module

lukpueh opened the pull request:
securesystemslib.hash is a small wrapper around hashlib, which serves two main purposes:
* provide helper function to hash a file
* translate custom hash algorithm name "blake2b-256" to "blake2b" with (digest_size=32).

In preparation for the removal of securesystemslib.hash, this patch ports above behavior to tuf and uses the builtin hashlib directly where possible.

related secure-systems-lab/securesystemslib#943

Commit:
Port securesystemslib.hash module

securesystemslib.hash is a small wrapper around hashlib, which serves
two main purposes:
* provide helper function to hash a file
* translate custom hash algorithm name "blake2b-256" to "blake2b" with
  (digest_size=32).

In preparation for the removal of securesystemslib.hash, this patch ports
above behavior to tuf and uses the builtin hashlib directly where
possible.

related secure-systems-lab/securesystemslib#943

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

lukpueh commented on tuf/api/_payload.py:
```diff
@@ -45,6 +48,31 @@
 T = TypeVar("T", "Root", "Timestamp", "Snapshot", "Targets")
 
 
+def _hash(algo: str) -> Any:  # noqa: ANN401
```
Python's hashlib does not make the `Hash` type public, hence I'm using `Any` and ask the linter for forgiveness.

lukpueh commented on tuf/api/_payload.py:
```diff
@@ -664,19 +692,15 @@ def _verify_hashes(
         data: bytes | IO[bytes], expected_hashes: dict[str, str]
     ) -> None:
         """Verify that the hash of ``data`` matches ``expected_hashes``."""
-        is_bytes = isinstance(data, bytes)
```
Moving this further down appeases mypy.

jku replied:
theoretically less efficient but seems like a fine compromise

lukpueh replied:
Yes, it's a shame. Maybe there's a better way to narrow the type here, but I couldn't find one. Let's come back to the performance question, if it becomes an issue.

lukpueh commented on tests/repository_simulator.py:
```diff
@@ -80,6 +80,8 @@
 
 SPEC_VER = ".".join(SPECIFICATION_VERSION)
 
+_DEFAULT_HASH_ALGORITHM = "sha256"
```
IMO it's okay to make this a local default, and not re-use the one newly added to _payload.py, so that we can keep the latter internal.

jku replied:
sure, I'd be fine with just a magic value as well: it's not really even a default, it's just what this repository decided to use

lukpueh commented on tuf/api/_payload.py:
```diff
@@ -45,6 +48,31 @@
+    f.seek(0)
+    try:
+        # Only availble on Pythons 3.11+
+        digest = hashlib.file_digest(f, lambda: _hash(algo))  # type: ignore[arg-type]
```
Hm. mypy running on ci, thinks this ignore is unnecessary, probably because it runs on a version of Python <3.11. :(

lukpueh commented on tuf/api/_payload.py:
```diff
@@ -45,6 +49,30 @@
+    """Returns hashed file."""
+    f.seek(0)
+    if sys.version_info >= (3, 11):
+        digest = hashlib.file_digest(f, lambda: _hash(algo))  # type: ignore[arg-type]
```
Mypy does not allow `IO[bytes]` for f here. This may be related to https://github.com/python/typing/issues/659. Ignore seems fair.

jku commented on tuf/api/_payload.py:
```diff
@@ -731,21 +755,16 @@ def _get_length_and_hashes(
-                        data, algorithm
-                    )
-            except (
-                sslib_exceptions.UnsupportedAlgorithmError,
-                sslib_exceptions.FormatError,
-            ) as e:
+                    digest_object = _file_hash(data, algorithm)
+            except (ValueError, TypeError) as e:
```
the different functionality is a little annoying here (you need to call update on one digest_object but not the other)... 
* If the `_hash` and `_file_hash` functions directly returned the hexdigest this code would be simpler and the functions would be properly annotated
* on the other hand then _hash() would need a data argument (meaning you would have to split the lambda needed by hashlib.file_digest() into another function)

I don't really know what is the simplest... up to you

lukpueh replied:
Does [535a189](https://github.com/theupdateframework/python-tuf/pull/2815/commits/535a18918bff8896642ac13602aa3c53e3bdb14b) look better? I don't have very strong feelings about either way.

jku replied:
No passionate argument here either.  I think I like this new version better.

jku reviewed (approved):
looks fine to me. Left some comments

jku commented:
> coverage: 96.981% (-0.1%) from 97.076%

coveralls.io is down again so I'm not sure... but I'm guessing this is the "blake2b-256" special case -- otherwise we're probably exercising this code.

Commit:
Rename hash algo global in repo simulator

Remove the "default" prefix, because it's not a default but rather a
fixed value.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Refactor hash helpers

Consolidate interface of bytes hash and file hash helpers.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

lukpueh commented:
> > coverage: 96.981% (-0.1%) from 97.076%
> 
> coveralls.io is down again so I'm not sure... but I'm guessing this is the "blake2b-256" special case -- otherwise we're probably exercising this code.

Yep, blake is the culprit. Let me add a test.

Commit:
Add tests for custom blake hash

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

lukpueh commented:
@jku, want to re-approve after my latest changes?

jku reviewed (approved):
Looks good. The only nit I have left is that multiple docstrings still refer to _"securesystemslib default hash algorithm"_.

Commit:
docs: Remove reference to securesystemslib hash

Default hash sha256 is now defined locally.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

lukpueh commented:
> Looks good. The only nit I have left is that multiple docstrings still refer to "securesystemslib default hash algorithm".

Good eyes! I pushed a fix and will merge when ci returns.
